Deploy perf guide (#1)
noahgibbs/rails: guides/source/deploy_performance_tuning.md

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**

Tuning Performance for Deployment
=================================

This guide covers performance and concurrency configuration for deploying your production Ruby on Rails application.

After reading this guide, you will know:

* Whether to use Puma, the default application server
* How to configure important performance settings for Puma

More information about how to configure your application can be found in the [Configuration Guide](configuring.html).

--------------------------------------------------------------------------------

Choosing an Application Server
------------------------------

> An application server uses a particular concurrency method. For example Unicorn uses processes, Puma is hybrid process- and thread-based concurrency, Thin uses EventMachine and Falcon uses Ruby Fibers.

Reviewer: ```suggestion
An application server uses a particular concurrency method. For example Unicorn uses processes, Puma is hybrid process- and thread-based concurrency and Falcon uses Ruby Fibers.
```

Not sure if it's still worth mentioning Thin today. However it might make sense to mention Passenger: https://rails-hosting.com/2022/#which-rails-servers-are-you-using-in-production

Author: I mostly mention Thin because it's mentioned in other Guides, so I know it's okay to mention, and because it has an unusual concurrency method (like Falcon for Fibers). But yeah, not common in production. I've left Thin, but included Passenger.

> A full discussion of Ruby's concurrency methods is beyond the scope of this document. If you want to use a method other than processes or threads, you will need to use a different application server. There are other features are only available with another server, such as Pitchfork's preforking.

Reviewer: ```suggestion
A full discussion of Ruby's concurrency methods is beyond the scope of this document. If you want to use a method other than processes or threads, you will need to use a different application server. There are other features are only available with another server, such as Pitchfork's reforking.
```

But honestly I'm not sure `Pitchfork` is worth a specific mention here.

The most common application servers are used by removing the Puma gem from your Gemfile and including the gem for the server. Consult the appropriate application server documentation for details.

Ruby Concurrency
----------------

Ruby has many kinds of concurrency. Puma supports a hybrid process-based and thread-based concurrency model.

> This guide is for [CRuby](https://ruby-lang.org), the original implementation of Ruby. If you're using a Ruby without a GVL such as JRuby or TruffleRuby, the configuration will be very different. If needed, check other sources specific to your Ruby implementation.

Reviewer: ```suggestion
This guide assumes you are running [CRuby](https://ruby-lang.org), the canonical implementation of Ruby. If you're using a Ruby implementation without a GVL nor forking support, such as JRuby or TruffleRuby, most of this guide doesn't apply, and you likely want to run a single Puma process with a large number of threads. If needed, check other sources specific to your Ruby implementation.
```

Author: Okay. Added this with minor tweaks.

### Process-Based Concurrency

Puma calls its multi-process concurrency "clustered mode". In this mode it forks new worker processes from a master process and each one separately processes requests. Each worker is a fully-capable Ruby process, doing everything Ruby normally does.

> Creating new processes uses a lot of memory. Each worker contains all data from the master process. Ruby uses [copy-on-write memory](https://en.wikipedia.org/wiki/Copy-on-write) to avoid duplicating most master-process data that doesn't change. But process-based workers often use a lot of memory, especially long-running workers.

Reviewer: ```suggestion
Each forked worker significantly increase memory usage since each worker contains all data from the parent process. But Ruby leverages [copy-on-write memory](https://en.wikipedia.org/wiki/Copy-on-write) to avoid duplicating most master-process data that doesn't change. But process-based workers often use a lot of memory, especially long-running workers.
```

> Processes are resilient. Killing a single process doesn't affect other processes much. They are also slower to start up than threads. Loading your application in the master thread instead of the workers is called preloading. It helps startup time, but may increase memory use.

Reviewer: ```suggestion
Processes are resilient. Killing or crashing a single process doesn't affect other processes at all. Loading your application in the master process instead of the workers is called preloading. It helps reduce memory usage by increasing the amount of memory that can be shared with the parent via Copy-on-Write.
```

Author: Incorporated with minor tweaks.

### Thread-Based Concurrency

> Multiple threads can run in the same process. This avoids multiple copies of shared data. Thread-based workers usually use much less memory than the same number of process-based workers.

Reviewer: ```suggestion
Multiple threads can run in the same process. This avoids multiple copies of shared data. Thread-based workers usually use much less memory than the same number of process-based workers.

However they tend to cause memory fragmentation in the heap, as such if running multiple threads it is heavily recommended to use jemalloc or configure glibc with `MALLOC_ARENA_MAX =1`
```

Author: I didn't mention memory config in this section, but added its own section and mentioned it in testing.

> [CRuby](https://www.ruby-lang.org/en/) has a [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), often called the GVL or GIL. The GVL prevents multiple threads from running Ruby code at the same time in a single process. A thread can be waiting on network data, database operations or some other non-Ruby work, but only one can actively run Ruby code at a time. This means thread-based concurrency is more efficient for applications that use a lot of I/O such as database operations or network APIs. The more I/O your application uses, the more threads it would benefit from.

Reviewer: ```suggestion
[CRuby](https://www.ruby-lang.org/en/) has a [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), often called the GVL or GIL. The GVL prevents multiple threads from running Ruby code at the same time in a single process. Multiple threads can be waiting on network data, database operations or some other non-Ruby work, but only one can actively run Ruby code at a time. This means thread-based concurrency is more efficient for applications that use a lot of I/O such as database operations or network APIs. The more I/O your application uses, the more threads it would benefit from.
```

> [CRuby](https://www.ruby-lang.org/en/) has a [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), often called the GVL or GIL. The GVL prevents multiple threads from running Ruby code at the same time in a single process. A thread can be waiting on network data, database operations or some other non-Ruby work, but only one can actively run Ruby code at a time. This means thread-based concurrency is more efficient for applications that use a lot of I/O such as database operations or network APIs. The more I/O your application uses, the more threads it would benefit from.

Reviewer: Similarly, the Ruby garbage collector is "stop the world", when it triggers, all the threads are stopped. So one request allocating a lot of objects can impact the performance of other threads.

Author: Good idea. Mentioned.

> With the GVL, using a lot of threads has limited value. A Rails app rarely benefits from more than 6. To have a large number of workers, some other concurrency method should be used.

Reviewer: ```suggestion
With the GVL, using a lot of threads has diminishing returns. A Rails app rarely benefits from more than 6. To have a large number of workers, some other concurrency method should be used.
```

Threads are less resilient than processes. Certain errors like segmentation faults can destroy the entire process and all threads inside.

> Thread-based workers have much faster startup time than process-based workers. This isn't an issue for most long-running Rails servers, but can be important in some cases.

Reviewer: Not sure the startup time is all that different. Assuming `preload_app!`, the whole boot sequence is the same, then it's `Process.fork` vs `Thread.new`, one is indeed slower, but we're talking milliseconds here, so I'd probably scrap this consideration altoghether to not overload the reader.

Author: Yeah, I think you're right. I'll remove references to startup time and add a bit more about memory allocator, which you rightly bring up as an important issue.

### Hybrid Concurrency

Puma allows forking multiple processes, each of which uses multiple threads. This provides a compromise between process-based and thread-based concurrency. Using multiple threads per process helps memory usage. Multiple processes permit running more Ruby code at the same time since there is one GVL per process.

Hybrid concurrency limits the damage from a segmentation fault or other error that kills a process. A single process dying will kill that process's threads but not threads in other processes.

Choosing Default Settings
-------------------------

> Rails' default settings are chosen for small applications. You can improve performance for your large application that serves a lot of requests by changing settings.

Reviewer: I'm not sure it's particularly true.

Author: I was thinking in terms of, say, increasing threads per process from 3 or increasing the number of processes. But I could make the processes part a lot clearer in the guide.

This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.

> If you're using a Ruby implementation without a GVL such as JRuby or TruffleRuby the number of threads and processes will be completely different. These recommendations are only accurate for CRuby. These are also production recommendations. Your development application will have different needs from a production application, and should use a different configuration.

Reviewer: ```suggestion
These are production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
```

Already said in the intro so we can probably ✂️

Author: Good catch, thanks!

### Threads Per Process

> Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Small applications with less I/O benefit from around 3 threads per process.

Reviewer: ```suggestion
The default Puma configuration generated by Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Small applications with less I/O benefit from around 3 threads per process.
```

Reviewer: > A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum.

I think the key is to define IO heavy, as not everyone might have the same definition of heavy.

Reviewer: Also I don't think it's really a matter of application size.

Author: Fair. I've expanded this a bit to mention databases, Redis and network APIs. I think it illustrates the point better. Good suggestion!

From the default Puma configuration:

> As a rule of thumb, increasing the number of threads will increase how much
> traffic a given process can handle (throughput), but due to CRuby's
> Global VM Lock (GVL) it has diminishing returns and will degrade the
> response time (latency) of the application.

To set the number of threads, you can change the call to the +threads+ method in +config/puma.rb+. Or you can set the +RAILS_MAX_THREADS+ environment variable, which will do the same.

### Number of Processes

> [Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) recommends running at least 1.5 processes per available processor core. Automatic methods to determine the number of cores are unreliable. You should specify the number of processes manually.

Reviewer: Note that 1.5 per core is if you run in pure process mode. If you run threads on top, then it's 1 per core.

I can't find where but I mentioned that to Nate and he said he would update that doc.

Author: Sounds good. That makes more sense to me, but I was worried I was missing something. Fixed. Thanks!

To set the number of worker processes, you can change the call to the +workers+ method in +config/puma.rb+. Or you can set the +WEB_CONCURRENCY+ environment variable, which will do the same.

### Preloading

> Puma creates new workers from a master process. By loading your application code in the master process, you can avoid doing so after creating the worker. This allows faster startup time.

Reviewer: ```suggestion
Puma creates new workers from a master process. By loading your application code in the master process, you can avoid doing so after creating the worker. This allows faster startup time, and reduce memory usage via Copy-on-Write.
```

### What to Measure

Throughput is the number of requests per second that your application successfully processes. Any good load testing program will measure it. Throughput is normally a single number for each load test.

Latency is the delay from the time the request is sent until its response is successfully received. Each individual request will have its own latency.

[Percentile](https://en.wikipedia.org/wiki/Percentile_rank) latency gives the latency where a certain percentage of requests have better latency than that. For instance, P90 is the 90th-percentile latency. The P90 is the latency for a single load test where only 10% of requests look longer than that to process. The P50 is the latency such that half your requests were slower, also called the median latency.

"Tail latency" refers to high-percentile latencies. For instance, the P99 is the latency such that only 1% of your requests was worse. P99 is a tail latency. P50 is not a tail latency.

### What You Can Change

> You can change the number of processes or threads in your test. You can also change other Puma configuration options such as wait_for_less_busy_worker, though you don't normally need to change them.

Reviewer: ```suggestion
You can change the number of threads in your test.
```

Threads are pretty much the only meaningful knob to turn. Which is good because multi-dimenstional testing is a pain.

Author: I've clarified when and how you may want to vary number of processes and mentioned memory testing.
